### DISCUSSION TITLE
PUT / POST Request to API

### DISCUSSION THREAD

**Discussion by florianschendel** [netbox-community/netbox#5915] [General]
Hi,

I have some problems with modifying the existing objects in Netbox with PUT or PATCH via API.
With the PATCH command I get always a result of 200, I see in the log there was a change on object but no value was updated. With the PUT statement I get the error code 400 with the message.

```
{'vid': ['This field is required.'], 'name': ['This field is required.']}
```
I tried 2.10.3 + the latest 2.10.5 version.

Here is my python sample

```
apiBaseUrl = "http://" + netboxServerHostname + ":8000" + "/api"
headers = {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    'Authorization': "Token " + authToken
}

vlan_json = {
'vid': '1234',
'name': "test",
'description': "test",
'tenant': '1',
'region': '1'
}

patch_json = json.dumps(vlan_json)
patchResult = requests.put(apiBaseUrl + '/ipam/vlans/' + str(json_id),headers=headers,verify=False,data=patch_json)

```

Any ideas?

**Comment by candlerb** [comment]
This was fun :-)

The first things I spotted (which are not the main problem):
* `vid` should be a number not a string
* `tenant`/`region` should also be a number not a string, or else a nested serializer like `{id: <number>}` - see [here](https://netbox.readthedocs.io/en/stable/rest-api/overview/#related-objects)
* you can avoid the json.dumps step, by passing `json=vlan_json` instead of `data=patch_json` to requests.put/patch.

I first replicated your problem like this:

```
curl netbox/api/ipam/vlans/8/ >z1
curl -H "Authorization: Token XXXXXXXX" -XPATCH -d '{"vid":1234}' netbox/api/ipam/vlans/8/
curl netbox/api/ipam/vlans/8/ >z2
diff -u z1 z2
```

What I found was that the `last_updated` field changed, but not the `vid`.  However, it worked fine if I also set `-H "Content-Type: application/json"`.  This to me seems like a bug in Netbox (the PATCH should have been rejected if the Content-Type was not acceptable).  However, your python *is* setting this header, so that's the not problem.

Then I replicated with your python code, and I found the same as you.  I did a tcpdump, and I was able to see the headers were *not* being set.  Not only that, but there was no data in the body!

```
PATCH /api/ipam/vlans/8/ HTTP/1.1
Host: netbox.example.net
Content-Length: 0
User-Agent: python-requests/2.25.1
Accept-Encoding: gzip, deflate
Accept: application/json
Connection: keep-alive
Authorization: Token XXXXXXXX
```

But tcpdump also showed what the real problem was:

```
HTTP/1.1 301 Moved Permanently       <<<<<
Date: Wed, 03 Mar 2021 08:31:51 GMT
Server: gunicorn/20.0.4
Content-Type: text/html; charset=utf-8
Location: /api/ipam/vlans/8/         <<<<<
Content-Length: 0
Vary: Origin
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
```

You had forgotten the trailing slash from the URL.  Netbox issued a redirect, and requests followed the redirect; but when it re-issued the PATCH to the new URL, the body was empty and the headers forgotten.  This seems like a bug in requests to me. 

Anyway, here is the fixed version:

```
headers = {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    'Authorization': "Token " + authToken
}

vlan_json = {
'vid': 1234,
}

patchResult = requests.patch(apiBaseUrl + '/ipam/vlans/' + str(json_id) + '/',headers=headers,json=vlan_json)
print(patchResult.text)
```

I suggest it might be better to use pynetbox instead of raw requests going forward, as it wraps all this nicely for you.

**Comment by florianschendel** [comment]
I created also a Wireshark dump to verify PUT request

Here the Screenshoot
![Bildschirmfoto 2021-03-03 um 09 37 11](https://user-images.githubusercontent.com/33891080/109777578-2ce74700-7c04-11eb-8a59-6cec1f5ef8a4.png)

Full Wireshark dump.
[PUT-Request.pcapng.zip](https://github.com/netbox-community/netbox/files/6074607/PUT-Request.pcapng.zip)

**Reply by candlerb** [reply]
I think we crossed over :-)

If you look carefully at your wireshark you will see:
* `PUT /api/ipam/vlans/1` with correct headers and body
* Netbox responds with `HTTP/1.1 301 Moved Permanently` (redirect)
* `PUT /api/ipam/vlans/1/` again.  In my case I saw missing Content-Type header and empty body, but you might have a different version of `requests` which behaves differently
* Netbox responds with 400 Bad Request (whereas I got 200 OK).  You'll need to look at the response body to see what it objects to. It might be to do with using strings instead of numbers for vid/region/tenant.

**Reply by florianschendel** [reply]
> I think we crossed over :-)
> 
> If you look carefully at your wireshark you will see:
> 
> * `PUT /api/ipam/vlans/1` with correct headers and body
> * Netbox responds with `HTTP/1.1 301 Moved Permanently` (redirect)
> * `PUT /api/ipam/vlans/1/` again.  In my case I saw missing Content-Type header and empty body, but you might have a different version of `requests` which behaves differently
> * Netbox responds with 400 Bad Request (whereas I got 200 OK).  You'll need to look at the response body to see what it objects to. It might be to do with using strings instead of numbers for vid/region/tenant.

Yes! Thanks a lot for your EYES! :-) It works now

**Reply by candlerb** [reply]
As for the difference between yours and mine: the wireshark dump shows you were doing PUT rather than PATCH, so the 400 error said

```
{"vid":["This field is required."],"name":["This field is required."]}
```

given the empty body to PUT.

**Comment by eliasbats** [comment]
You can't imagine how much pain that forgotten trailing slash gave me... Thanks so much for this...
